misc/ls-extensions/modules/performance-analyzer-services/src/main/java/io/ballerina/PerformanceAnalyzerNodeVisitor.java

```diff
@@ -0,0 +1,536 @@
+    }
+
+    @Override
+    public void visit(ModuleVariableDeclarationNode moduleVariableDeclarationNode) {
+
+        Optional<Symbol> symbol = model.symbol(moduleVariableDeclarationNode);
+        if (symbol.isEmpty() || moduleVariableDeclarationNode.initializer().isEmpty()) {
+            return;
+        }
+
+        boolean clientObject = isClientObject(symbol.get());
+        if (clientObject) {
+            resolveEndPoint(symbol.get(), moduleVariableDeclarationNode.initializer().get());
+        }
+
+        LineRange lineRange = moduleVariableDeclarationNode.typedBindingPattern().bindingPattern().
+                location().lineRange();
+
+        if (clientObject || isRecordObject(symbol.get())) {
+            Optional<ExpressionNode> expressionNode = moduleVariableDeclarationNode.initializer();
+
+            if (expressionNode.isPresent() && expressionNode.get().kind() != SyntaxKind.SIMPLE_NAME_REFERENCE) {
+                this.registerVariableRef(lineRange, moduleVariableDeclarationNode);
+            }
+        } else if ((symbol.get().kind() == SymbolKind.VARIABLE) &&
+                (moduleVariableDeclarationNode.initializer().get().kind() == SyntaxKind.REMOTE_METHOD_CALL_ACTION)) {
+
+            this.registerVariableRef(lineRange, moduleVariableDeclarationNode);
+        }
+
+        if (moduleVariableDeclarationNode.typedBindingPattern().bindingPattern().kind() ==
+                SyntaxKind.CAPTURE_BINDING_PATTERN) {
+
+            SyntaxKind initializerKind = moduleVariableDeclarationNode.initializer().get().kind();
+            if (initializerKind == SyntaxKind.NUMERIC_LITERAL || initializerKind == SyntaxKind.STRING_LITERAL) {
+                this.registerVariableRef(lineRange, moduleVariableDeclarationNode);
+            }
+        }
+        moduleVariableDeclarationNode.typedBindingPattern().accept(this);
+        moduleVariableDeclarationNode.initializer().ifPresent(initializer -> initializer.accept(this));
+    }
+
+    @Override
+    public void visit(FunctionDefinitionNode functionDefinitionNode) {
+
+        LineRange lineRange = functionDefinitionNode.lineRange();
+        if (functionDefinitionNode.syntaxTree().filePath().equals(file)
+                && range.getStart().getLine() == lineRange.startLine().line()
+                && range.getStart().getCharacter() == lineRange.startLine().offset()
+                && range.getEnd().getLine() == lineRange.endLine().line()
+                && range.getEnd().getCharacter() == lineRange.endLine().offset()
+        ) {
+            withinRange = true;
+        }
+
+        functionDefinitionNode.functionBody().accept(this);
+        withinRange = false;
+    }
+
+    @Override
+    public void visit(VariableDeclarationNode variableDeclarationNode) {
+
+        Optional<Symbol> symbol = model.symbol(variableDeclarationNode);
+        if (symbol.isEmpty() || variableDeclarationNode.initializer().isEmpty()) {
```

Comment:
If you do the `variableDeclarationNode.initializer().isEmpty()` before accessing the `SemanticModel.symbol` API, we can save performance cycles. Please check